### PR TITLE
Enum Functions

### LINKED ISSUE #2740: Question: Can one easily determine if a string value matches the label of an enum?
Hello,

In Postgres one can apparently determine if a string value matches an enum label by using the **enum_range** function in combination with **any** and casting the returned array to **name[]**:   

SELECT enum_range(null::known_roles_list);
  enum_range
---------------
 {role1,role2}
 
SELECT current_user = **any**(enum_range(null::known_roles_list)::**name[]**)

source: https://stackoverflow.com/questions/51924066/postgresql-check-if-value-is-equal-to-an-enum-value

How can we check for a match in duckdb?  Is an attempted cast to the enum type the way to accomplish this? 

Regards,
Chris

### LINKED ISSUE #2741: Add Enum Support Functions
Reference: https://www.postgresql.org/docs/14/functions-enum.html
```sql 
CREATE TYPE rainbow AS ENUM ('red', 'orange', 'yellow', 'green', 'blue', 'purple');
```

- [ ] enum_first ( anyenum ) → anyenum
Returns the first value of the input enum type.
e.g., enum_first(null::rainbow) → red

- [ ] enum_last ( anyenum ) → anyenum
Returns the last value of the input enum type.
e.g., enum_last(null::rainbow) → purple

- [ ] enum_range ( anyenum ) → anyarray
Returns all values of the input enum type in an ordered array.
e.g., enum_range(null::rainbow) → {red,orange,yellow,​green,blue,purple}

- [ ] enum_range ( anyenum, anyenum ) → anyarray
Returns the range between the two given enum values, as an ordered array. The values must be from the same enum type. If the first parameter is null, the result will start with the first value of the enum type. If the second parameter is null, the result will end with the last value of the enum type.
e.g., enum_range('orange'::rainbow, 'green'::rainbow) → {orange,yellow,green}
enum_range(NULL, 'green'::rainbow) → {red,orange,​yellow,green}
enum_range('orange'::rainbow, NULL) → {orange,yellow,green,​blue,purple}

### PR DESCRIPTION
**Pull request by pdet**
Issue: #2741 and #2740

### CODE REVIEW

File: src/function/scalar/enum/enum_functions_implementation.cpp [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+	if (arguments[0]->return_type.id() != LogicalTypeId::ENUM) {
+		throw std::runtime_error("This function needs an ENUM as an argument");
+	}
+	return make_unique<CurrentBindData>(context);
+}
+
+unique_ptr<FunctionData> BindEnumRangeBoundaryFunction(ClientContext &context, ScalarFunction &bound_function,
+                                                       vector<unique_ptr<Expression>> &arguments) {
+	if (arguments[0]->return_type.id() != LogicalTypeId::ENUM && arguments[0]->return_type != LogicalType::SQLNULL) {
+		throw std::runtime_error("This function needs an ENUM as an argument");
```
Comment (Mytherin) [inline src/function/scalar/enum/enum_functions_implementation.cpp:85]: Can we add tests that cover all of these exceptions?

File: src/common/types.cpp [hunk 2]
```diff
@@ -1150,6 +1150,19 @@ const string EnumType::GetValue(const Value &val) {
 	}
 }
 
+idx_t EnumType::GetValuePosition(const Value &val) {
+	switch (val.type().InternalType()) {
+	case PhysicalType::UINT8:
+		return val.value_.utinyint;
```
Comment (Mytherin) [inline src/common/types.cpp:1156]: `val.GetValue<uint32_t>()` should cover all these three cases so the switch is no longer required

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good. Some comments:

File: src/common/types.cpp (hunk 2 above)
Comment (pdet) [reply src/common/types.cpp:1156]: it doesn't but i've moved it there.

File: src/function/function.cpp [hunk 3]
```diff
@@ -27,7 +27,8 @@ FunctionData::~FunctionData() {
 }
 
 unique_ptr<FunctionData> FunctionData::Copy() {
-	throw InternalException("Unimplemented copy for FunctionData");
```
Comment (hawkfish) [inline src/function/function.cpp:30]: Why was this removed? I've personally found it quite helpful!

File: src/function/scalar/enum/enum_functions_implementation.cpp [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+	auto val = Value::LIST(enum_values);
+	result.Reference(val);
+}
+
+unique_ptr<FunctionData> BindEnumFunction(ClientContext &context, ScalarFunction &bound_function,
+                                          vector<unique_ptr<Expression>> &arguments) {
+	if (arguments[0]->return_type.id() != LogicalTypeId::ENUM) {
+		throw BinderException("This function needs an ENUM as an argument");
+	}
+	return make_unique<FunctionData>();
```
Comment (hawkfish) [inline src/function/scalar/enum/enum_functions_implementation.cpp:64]: Just return `nullptr`?

File: src/function/scalar/enum/enum_functions_implementation.cpp [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+	if (arguments[0]->return_type == LogicalType::SQLNULL && arguments[1]->return_type == LogicalType::SQLNULL) {
+		throw BinderException("This function needs an ENUM as an argument");
+	}
+	if (arguments[0]->return_type.id() == LogicalTypeId::ENUM &&
+	    arguments[1]->return_type.id() == LogicalTypeId::ENUM &&
+	    arguments[0]->return_type != arguments[1]->return_type) {
+
+		throw BinderException("The parameters need to link to ONLY one enum OR be NULL ");
+	}
+	return make_unique<FunctionData>();
```
Comment (hawkfish) [inline src/function/scalar/enum/enum_functions_implementation.cpp:84]: Return `nullptr`?

File: src/function/scalar/enum/enum_functions_implementation.cpp [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+void EnumFirst::RegisterFunction(BuiltinFunctions &set) {
+	set.AddFunction(ScalarFunction("enum_first", {LogicalType::ANY}, LogicalType::VARCHAR, EnumFirstFunction, false,
```
Comment (hawkfish) [inline src/function/scalar/enum/enum_functions_implementation.cpp:88]: I know you have a binder, but maybe `ANY` should be `ENUM`? I'm not sure how the candidate functions will be displayed in the situation where there are too many or too few arguments.

File: test/sql/function/enum/test_enum_range.test [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+statement ok
+CREATE TYPE currency AS ENUM ('usd', 'brl', 'eur');
+
+query I
+SELECT enum_range(null::rainbow)
+----
+[red, orange, yellow, green, blue, purple]
+
+query I
+SELECT enum_range_boundary('orange'::rainbow, 'green'::rainbow)
```
Comment (hawkfish) [inline test/sql/function/enum/test_enum_range.test:20]: What happens if you ask for green, orange?

**Review by hawkfish — CHANGES_REQUESTED** [review]

File: src/function/scalar/enum/enum_functions_implementation.cpp (hunk 6 above)
Comment (pdet) [reply src/function/scalar/enum/enum_functions_implementation.cpp:88]: The problem here is that the creation and equality of enum types are associated with the enum categories. I guess we could to these matches on LogicalTypeIDs, but I think @Mytherin prefers these binding setups

**Review by hawkfish — APPROVED** [review]

File: src/function/scalar/enum/enum_functions_implementation.cpp (hunk 6 above)
Comment (Mytherin) [reply src/function/scalar/enum/enum_functions_implementation.cpp:93]: Binding on an empty type is how the list functions do this, e.g. unnest binds on `{LogicalTypeId::LIST}`. Perhaps this should indeed bind on `LogicalTypeId::ENUM` instead of `ANY`.

**Comment by Mytherin** [comment]
Thanks!
